### PR TITLE
C#: Add queries to check untrusted data flow to external APIs

### PR DESCRIPTION
**Pull request by tamasvajk**
This is a port of the Java matching checks. 

The main differences are that 
* we're not only reporting `Method`s but `Callable`s,
* `SafeExternalAPICallable` adds only a super minimal safe API list for the time being.

### CODE REVIEW

**Comment by tamasvajk** [comment]
[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/631/)

File: csharp/ql/src/Security Features/CWE-020/ExternalAPIsUsedWithUntrustedData.ql [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *              frequently the API is called, and how many unique sources of untrusted data flow
+ *              to it.
+ * @id csharp/count-untrusted-data-external-api
+ * @kind table
+ * @tags security external/cwe/cwe-20
+ */
+
+import csharp
+import semmle.code.csharp.security.dataflow.ExternalAPIs
+import semmle.code.csharp.dataflow.DataFlow
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-020/ExternalAPIsUsedWithUntrustedData.ql:13]: I don't think this import is needed.

File: csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * database.
+ */
+
+import csharp
+import semmle.code.csharp.dataflow.flowsources.Remote
+import semmle.code.csharp.dataflow.TaintTracking
+import semmle.code.csharp.frameworks.System
+
+/**
+ * A `Callable` that is considered a "safe" external API from a security perspective.
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll:12]: `` `Callable` `` -> `callable`

File: csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    ) and
+    // Defined outside the source archive
+    not call.getTarget().fromSource() and
+    // Not a call to a method which is overridden in source
+    not exists(Virtualizable m |
+      m.overridesOrImplementsOrEquals(call.getTarget().getSourceDeclaration()) and
+      m.fromSource()
+    ) and
+    // Not already modeled as a taint step
+    not exists(DataFlow::Node next | TaintTracking::localTaintStep(this, next)) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll:57]: This won't work in all cases. I suggest we instead add
```ql
private class SummarizedCallableSafe extends SafeExternalAPICallable {
  SummarizedCallableSafe() { this instanceof SummarizedCallable }
}
```

File: csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    // Not a call to a known safe external API
+    not call.getTarget().getSourceDeclaration() instanceof SafeExternalAPICallable
+  }
+
+  /** Gets the called API `Callable`. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll:62]: `` `Callable` `` -> `callable`.

File: csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+/** A node representing untrusted data being passed to an external API. */
+class UntrustedExternalAPIDataNode extends ExternalAPIDataNode {
+  UntrustedExternalAPIDataNode() { any(UntrustedDataToExternalAPIConfig c).hasFlow(_, this) }
+
+  /** Gets a source of untrusted data which is passed to this external API data node. */
+  DataFlow::Node getAnUntrustedSource() {
+    any(UntrustedDataToExternalAPIConfig c).hasFlow(result, this)
+  }
+}
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/security/dataflow/ExternalAPIs.qll:89]:
```suggestion
class UntrustedExternalAPIDataNode extends ExternalAPIDataNode {
  private UntrustedDataToExternalAPIConfig c;
  UntrustedExternalAPIDataNode() { c.hasFlow(_, this) }

  /** Gets a source of untrusted data which is passed to this external API data node. */
  DataFlow::Node getAnUntrustedSource() { c.hasFlow(result, this) }
}
```

**Review by hvitved — CHANGES_REQUESTED** [review]
A few minor comments, otherwise LGTM. Great to have this query in the toolbox.

File: csharp/ql/test/query-tests/Security Features/CWE-020/ExternalAPIsUsedWithUntrustedData.expected [hunk 6]
```diff
@@ -1,2 +1,3 @@
+| System.Collections.Specialized.NameValueCollection.get_Item(string) [qualifier] | 1 | 1 |
```
Comment (hvitved) [inline csharp/ql/test/query-tests/Security Features/CWE-020/ExternalAPIsUsedWithUntrustedData.expected:1]: Interesting; we should model this (like we model dictionaries). (Not on this PR.)

**Review by hvitved — DISMISSED** [review]

**Comment by tamasvajk** [comment]
I rebased and pushed an extra commit to fix the CI check failure.

**Review by hvitved — APPROVED** [review]

**Pull request merged**